VSRAD.Package/DebugVisualizer/SliceVisualizer/SliceCellStyling.cs

```diff
@@ -1,64 +1,116 @@
-                    var rel2 = _table.SelectedWatch.GetRelativeValue(e.RowIndex, e.ColumnIndex - SliceVisualizerTable.DataColumnOffset, word: 1);
-                    var brush1 = new SolidBrush(GetHeatmapColor(rel1));
-                    var brush2 = new SolidBrush(GetHeatmapColor(rel2));
-
-                    e.Graphics.FillRectangle(brush1, new Rectangle(e.CellBounds.Left, e.CellBounds.Top, e.CellBounds.Width / 2, e.CellBounds.Height));
-                    e.Graphics.FillRectangle(brush2, new Rectangle(e.CellBounds.Left + e.CellBounds.Width / 2, e.CellBounds.Top, e.CellBounds.Width / 2, e.CellBounds.Height));
-                    e.Paint(e.CellBounds, DataGridViewPaintParts.All & ~DataGridViewPaintParts.Background);
-                    e.Handled = true;
-                }
-                else
-                {
-                    var relValue = _table.SelectedWatch.GetRelativeValue(e.RowIndex, e.ColumnIndex - SliceVisualizerTable.DataColumnOffset);
-                    e.CellStyle.BackColor = GetHeatmapColor(relValue);
-                }
+                HandleHeatMap(e);
             }
             else
             {
+                var colorIndex = e.ColumnIndex % _table.GroupSize == 0
+                    ? _table.GroupSize - SliceVisualizerTable.DataColumnOffset
+                    : e.ColumnIndex % _table.GroupSize - SliceVisualizerTable.DataColumnOffset;
+                var bgColor = DataHighlightColors.GetFromColorString(_stylingOptions.BackgroundColors, (int)colorIndex);
+                e.CellStyle.BackColor = _fontAndColor.FontAndColorState.HighlightBackground[(int)bgColor];
                 e.CellStyle.ForeColor = _fontAndColor.FontAndColorState.HighlightForeground[(int)DataHighlightColor.None];
-                e.CellStyle.BackColor = _fontAndColor.FontAndColorState.HighlightBackground[(int)DataHighlightColor.None];
+            }
+        }
+
+        private void PaintColumnSeparators(int dataColumnIndex, DataGridViewCellPaintingEventArgs e)
+        {
+            int width;
+            SolidBrush color;
+
+            if ((_columnStyling[dataColumnIndex] & ColumnStates.HasHiddenColumnSeparator) != 0)
+            {
+                width = _appearance.SliceHiddenColumnSeparatorWidth;
+                color = _fontAndColor.FontAndColorState.SliceHiddenColumnSeparatorBrush;
+            }
+            else if ((_columnStyling[dataColumnIndex] & ColumnStates.HasSubgroupSeparator) != 0)
+            {
+                width = _appearance.SliceSubgroupSeparatorWidth;
+                color = _fontAndColor.FontAndColorState.SliceSubgroupSeparatorBrush;
+            }
+            else return;
+
+            // We doing force paint of _visible_ part of cell.
+            // Since we have frozen columns visible part of cell
+            // is not necessarily whole cell.
+            var r = e.CellBounds.Left > _table.RowHeadersWidth
```

Comment:
What happens if you remove this check? I assume it can only be true for the first displayed column, do we need the `+ 1` offset there? Same goes for `VSRAD.Package.DebugVisualizer.CellStyling`, if we can get rid of the condition on `e.CellBounds.Left`, I think we can share this snippet between the two classes instead of duplicating it.